app/services/nic_ip_required/base.rb

```diff
@@ -0,0 +1,65 @@
+module NicIpRequired
+  class Base
+    attr_reader :nic, :from_compute, :subnet, :field, :other_field
+
+    delegate :logger, :to => :Rails
+    delegate :host, :domain, :to => :nic
+
+    def initialize(opts = {})
+      @nic = opts.fetch(:nic)
+
+      # the compute resource can provide an IP, so don't validate yet
+      @from_compute = opts.fetch(:from_compute, true)
+    end
+
+    def required?
+      # if it's not managed there's nowhere to specify an IP anyway
+      return false unless nic_managed?
+
+      # if the CR will provide an IP, then don't validate yet
+      return false if compute_provides_ip?
+
+      [require_ip_for_dns?, require_ip_for_dhcp?, require_ip_for_token?].any?
+    end
+
+    def nic_managed?
+      nic.host_managed? && nic.managed? && nic.provision?
+    end
+
+    def require_ip_for_dns?
+      reverse_dns? || forward_dns?
+    end
+
+    def reverse_dns?
+      subnet.present? && subnet.dns_id.present?
+    end
+
+    def forward_dns?
+      domain.present? && domain.dns_id.present? && !other_ip_protocol_provides_ip?
+    end
+
+    def require_ip_for_dhcp?
+      subnet.present? && subnet.dhcp_id.present?
+    end
+
+    def require_ip_for_token?
+      tokens_enabled? && token_required? && !other_ip_protocol_provides_ip?
+    end
+
+    def tokens_enabled?
+      Setting[:token_duration] == 0
+    end
+
+    def token_required?
```

Comment:
This might be called "token or IP required", or "unattended controller used?", as it defines whether the host will need some identifier to access UnattendedController - either use a token, or require an IP address.